onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc

```diff
@@ -211,49 +352,49 @@ bool QDQPropagationTransformer::PropagateQBackward(Graph& graph) const {
       continue;
     }
 
-    do {
-      if (q_node.InputNodesBegin() == q_node.InputNodesEnd()) {
-        break;
-      }
-      Node& prev_node = *graph.GetNode(q_node.InputNodesBegin()->Index());
-      if (!optimizer_utils::CheckOutputEdges(graph, prev_node, 1)) break;
-      if (!CanNodePropagate(prev_node)) {
-        // Try canceling out DQ/Q pair
-        Node& dq_node = prev_node;
-        if (graph_utils::IsSupportedOptypeVersionAndDomain(dq_node, "DequantizeLinear", {10, 13}) &&
-            graph_utils::IsSupportedProvider(dq_node, GetCompatibleExecutionProviders()) &&
-            TryCancelOutDQQPair(graph, dq_node, q_node)) {
-          is_modified = true;
-        }
+    const auto edge_before_q = GetPreviousEdge(graph, q_node);
+    if (!edge_before_q) {
+      continue;
+    }
+
+    for (auto curr_edge = GetPreviousPropagationEdge(graph, *edge_before_q);
+         curr_edge.has_value();
+         curr_edge = GetPreviousPropagationEdge(graph, *curr_edge)) {
+      if (auto* src_node = curr_edge->GetNodeAtEnd(graph, ExtendedGraphEdge::End::Source);
+          src_node && QDQ::MatchDQNode(*src_node)) {
         break;
       }
 
-      SwapAdjacentNodes(graph, prev_node, q_node);
-      is_modified = true;
-    } while (true);
+      ORT_RETURN_IF_ERROR(InsertQDQPair(graph, *curr_edge,
+                                        *q_input_defs[QDQ::InputIndex::SCALE_ID],
+                                        *q_input_defs[QDQ::InputIndex::ZERO_POINT_ID],
+                                        logger));
+      modified = true;
+    }
   }
 
-  return is_modified;
+  return Status::OK();
 }
+}  // namespace
 
-Status QDQPropagationTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+Status QDQPropagationTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level,
+                                            const logging::Logger& logger) const {
   GraphViewer graph_viewer(graph);
-  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  const auto node_indices = gsl::make_span(graph_viewer.GetNodesInTopologicalOrder());
 
-  for (auto node_index : node_topology_list) {
+  for (auto node_index : node_indices) {
     auto* node_ptr = graph.GetNode(node_index);
     if (node_ptr == nullptr)
       continue;  // node removed as part of an earlier fusion
 
     ORT_RETURN_IF_ERROR(Recurse(*node_ptr, modified, graph_level, logger));
   }
 
-  while (PropagateQBackward(graph) || PropagateDQForward(graph)) {
-  }
+  const auto& compatible_eps = GetCompatibleExecutionProviders();
 
-  modified = true;
+  ORT_RETURN_IF_ERROR(PropagateQBackward(graph, node_indices, compatible_eps, logger, modified));
+  ORT_RETURN_IF_ERROR(PropagateDQForward(graph, node_indices, compatible_eps, logger, modified));
 
```

Comment:
If we're inserting pairs do we need a forward and backwards pass?

Reply:
DQ forward pass handles: QDQ -> (op(s) that can propagate) -> (op that cannot propagate)
Q backward pass handles: (op that cannot propagate) -> (op(s) that can propagate) -> QDQ
so I think both directions are needed, unless I'm missing something